src/charm.py

```diff
@@ -363,6 +364,31 @@ def _on_get_admin_password(self, event: ActionEvent) -> None:
         """Returns the password for the user as an action response."""
         event.set_results({"admin-password": self.app_data.get("admin_password")})
 
+    def _on_set_admin_password(self, event: ActionEvent) -> None:
+        """Set the password for the admin user."""
+        # only leader can write the new password into peer relation.
+        if not self.unit.is_leader():
+            event.fail("The action can be run only on juju leader.")
+            return
+
+        new_password = generate_password()
+        if "password" in event.params:
+            new_password = event.params["password"]
+
+        with MongoDBConnection(self.mongodb_config) as mongo:
+            try:
+                mongo.set_user_password("admin", new_password)
+            except NotReadyError:
+                event.fail(
+                    "Failed changing the password: Not all members healthy or finished initial sync."
+                )
+                return
+            except PyMongoError as e:
+                event.fail(f"Failed changing the password: error={e}")
```

Comment:
```suggestion
                event.fail(f"Failed changing the password: {e}")
```

Nit for consistency only